test_data/asm/vm_to_block.asm

```diff
@@ -0,0 +1,75 @@
+machine Binary(latch) {
+
+    degree 8;
+
+    function and x, y -> z {
+    }
+
+    function or x, y -> z {
+    }
+
+    constraints {
+        col fixed latch = [1]*;
+        col witness x;
+        col witness y;
+        col witness z;
+        col fixed P_FUNCTION = [0, 0, 0, 0, 1, 1, 1, 1] + [1]*;
+        col fixed P_X = [0, 0, 1, 1, 0, 0, 1, 1] + [1]*;
+        col fixed P_Y = [0, 1, 0, 1, 0, 1, 0, 1] + [1]*;
+        col fixed P_Z = [0, 0, 0, 1, 0, 1, 1, 1] + [1]*;
+        { function_id, x, y, z } in { P_FUNCTION, P_X, P_Y, P_Z };
```

Comment:
The way I understand `function_id`, it will always be set to the index of the function being called (starting from 0), right? And in order to actually compute something different depending on the function, I have to incorporate it into my constraints? 

Two comments for that:
- That's pretty magic. E.g., if I change the order in which functions are declared or assume 1-indexing, things break.
- For a larger number of functions, expressing that `function_id == 42` in polynomial language is actually quite complicated (you'd be spelling out the Lagrange polynomial). It would be more efficient to represent it as a logarithmic number of bits. I *think* that's what Polygon Miden is doing!

Reply:
> The way I understand `function_id`, it will always be set to the index of the function being called (starting from 0), right? And in order to actually compute something different depending on the function, I have to incorporate it into my constraints?

Correct, `function_id` is currently injected.

> Two comments for that:
> 
> * That's pretty magic. E.g., if I change the order in which functions are declared or assume 1-indexing, things break.
Good point. That's the way Hermez does it and I can see how it's fragile. I can add a way to make this user-defined at the function level.
> * For a larger number of functions, expressing that `function_id == 42` in polynomial language is actually quite complicated (you'd be spelling out the Lagrange polynomial). It would be more efficient to represent it as a logarithmic number of bits. I _think_ that's what Polygon Miden is doing!
Right, I actually had that thought as well at some point! In Hermez I think it didn't make a difference since the block machine uses lookup arguments and never checks `function_id` with polynomials. I didn't know Miden used bits, nice!

Let me have a think about this, it's a bit awkward to find a syntax which unifies block machines and vms.

Reply:
Ok here's what I came up with:

```
function foo<42> x, y -> z {
}
```

`function_id` is still injected but the value is user-defined.

Reply:
Nice, I like it! It could also be generalized later to cover the bit use-case if needed, for example by allowing a tuple of function IDs that can be specified as `function foo<(1, 0, 1, 1)>`.

Reply:
Added an [issue](https://github.com/powdr-labs/powdr/issues/452), feel free to amend it